Add support for the gpu modifier at HLT

fwyzard opened the pull request:
#### PR description:

When the `gpu` modifier is enabled, make the Pixel, ECAL and HCAL reconstruction running at HLT offloadable to a GPU.

#### PR validation:

Tested with
```bash
cmsDriver.py reHLT \
    --processName reHLT \
    -s HLT:@relval2021 \
    --conditions 120X_mcRun3_2021_realistic_v4 \
    --datatier GEN-SIM-DIGI-RAW \
    -n 10 \
    --eventcontent FEVTDEBUGHLT \
    --geometry DB:Extended \
    --era Run3 \
    --procModifiers gpu \
    --filein /store/relval/CMSSW_12_0_0_pre6/RelValTTbar_14TeV/GEN-SIM-DIGI-RAW/PU_120X_mcRun3_2021_realistic_v4_JIRA_129-v1/00000/79c06ed5-929b-4a57-a4f2-1ae90e6b38c5.root
```

With this change, the `customizeHLTforPatatrack()` customisation will be tested as part of the `.502`, `.512` and `.522` GPU workflows, and `customizeHLTforPatatrackTriplets()` with the `.506` workflow.

fwyzard committed:
Add support for the gpu modifier at HLT

When the gpu modifier is enabled, make the Pixel, ECAL and HCAL
reconstruction offloadable to a GPU.

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34978/24821

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34788

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

- HLTrigger/Configuration (**hlt**)


@Martin-Grunewald can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f5f5f9/17966/summary.html
**COMMIT**: 1cdd51f4257b6819d5c367a75d171ca0ed7252ab
**CMSSW**: CMSSW_12_1_X_2021-08-22-2300/slc7_amd64_gcc900
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34978/17966/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19735
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19729
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_1_X_2021-08-22-2300+f5f5f9/44934/triggerResults): no differences found


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000352
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000324
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-22-2300+f5f5f9/44937/triggerResults): no differences found

fwyzard commented:
Mhm, no, this didn't work as intended.
Looks like the `gpu` modifier is only applied to the RECO stesp, not the HLT step, so nothing happened.

fwyzard committed:
Make the Patatrack workflows apply the 'gpu' modifier to the HLT step

fwyzard commented:
enable gpu

fwyzard commented:
please test

fwyzard commented:
rebased on top of CMSSW_12_1_0_pre2, and added support for running the HLT step on GPU in runTheMatrix

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34978/25137

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34788

jordan-martins commented:
+1

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34978/25148

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34788

cmsbuild commented:
Pull request #34978 was updated. @jordan-martins, @chayanit, @bbilin, @wajidalikhan, @Martin-Grunewald, @AdrianoDee, @srimanob, @kskovpen can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f5f5f9/18436/summary.html
**COMMIT**: 918d200b274018b79edbc89c148bad1fba67df0d
**CMSSW**: CMSSW_12_1_X_2021-09-08-2300/slc7_amd64_gcc900
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34978/18436/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19735
 - DQMHistoTests: Total failures: 72
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19663
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_1_X_2021-09-08-2300+f5f5f9/45303/triggerResults): no differences found


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3001001
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000979
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-08-2300+f5f5f9/45309/triggerResults): no differences found

fwyzard commented:
Looks OK as far as I can check remotely.

Martin-Grunewald commented on HLTrigger/Configuration/python/customizeHLTforCMSSW.py:
```diff
@@ -147,6 +150,10 @@ def customiseFor2018Input(process):
     
+    # if the gpu modifier is enabled, make the Pixel, ECAL and HCAL reconstruction offloadable to a GPU
+    from HLTrigger.Configuration.customizeHLTforPatatrack import customizeHLTforPatatrack
+    gpu.makeProcessModifier(customizeHLTforPatatrack).apply(process)
```
So on non-gpu cases, gpu as python variable exists but makeProcessModifier does nothing?

fwyzard replied:
`gpu` exists (it's imported above, at line 4), but it will be disabled if it's not used by the process (like an Era that is not used): in this case the customisation associated to `makeProcessModifier()` is not called.

missirol replied:
Hi @fwyzard , just wanted to ask a question, for my own education.

I was wondering if one can incur in the case where `customizeHLTforPatatrack` is run twice, now from `customizeHLTforPatatrack` plus once more if the user specifies something like `hltGetConfiguration --customise [..].customizeHLTforPatatrack`. I'm guessing this can't happen because one would not have a way to enable the `gpu` modifier from `hltGetConfiguration`. Or am I missing something? And do you already know if the Patatrack customisations are such that, even if run multiple times, they'd still return the correct configuration?

fwyzard replied:
hi Marino,
I have not tired, but based on its implementation I think that trying to apply the `customizeHLTforPatatrack` twice will not work as intended, and/or produce a broken configuration.

Specifying the `gpu` modifier and the `customizeHLTforPatatrack` at the same time would indeed be a mistake - like trying to apply the `customizeHLTforPatatrack` twice.

However, I don't think `hltGetConfiguration` supports process modifiers, so that shouldn't happen.

missirol replied:
Okay. Thanks for the info!

srimanob commented on Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:
```diff
@@ -415,10 +415,12 @@ def condition(self, fragment, stepList, key, hasHarvest):
         super(PatatrackWorkflow, self).__init__(
             steps = [
+                'Digi',
+                'DigiTrigger',
```
Adding DigiTrigger here means you expect also Phase-2?

Martin-Grunewald commented:
@fwyzard 
Is this and the bp PR now ready to go?

fwyzard commented:
I think so
I can double check things tomorrow, if needed.

fwyzard committed:
Simplify the GPU pixel local reconstruction

Simplify the Pixel local reconstruction running on GPU, taking advatage of #34259.

fwyzard commented:
@VinInn do the changes to the pixel local reconstruction sequence make sense to you ?

fwyzard commented:
please test

fwyzard commented:
AFAICT we still need the pixel digis to check for inactive or bad pixels in the `PixelInactiveAreaTrackingRegionsSeedingLayersProducer` modules (unless or until they are removed from the Run-3 menu)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34978/25300

- This PR adds an extra 36KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34788, #35269
  - File  HLTrigger/Configuration/python/customizeHLTforPatatrack.py  modified in PR(s): #35117

cmsbuild commented:
Pull request #34978 was updated. @jordan-martins, @chayanit, @bbilin, @wajidalikhan, @Martin-Grunewald, @AdrianoDee, @srimanob, @kskovpen can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f5f5f9/18636/summary.html
**COMMIT**: fdc1e737da34db68b105b41b9978d0e54c9ca7e2
**CMSSW**: CMSSW_12_1_X_2021-09-14-1100/slc7_amd64_gcc900
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34978/18636/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19735
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19729
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_1_X_2021-09-14-1100+f5f5f9/45474/triggerResults): no differences found


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000833
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000805
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-14-1100+f5f5f9/45478/triggerResults): no differences found

fwyzard commented:
rebased on top of CMSSW_12_1_0_pre3

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34978/25355

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34788, #35269, #35309
  - File  HLTrigger/Configuration/python/customizeHLTforPatatrack.py  modified in PR(s): #35117

cmsbuild commented:
Pull request #34978 was updated. @jordan-martins, @chayanit, @bbilin, @wajidalikhan, @Martin-Grunewald, @AdrianoDee, @srimanob, @kskovpen can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f5f5f9/18699/summary.html
**COMMIT**: dd84d38b0003198745bf74a832b24af46c5f817d
**CMSSW**: CMSSW_12_1_X_2021-09-16-2300/slc7_amd64_gcc900
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34978/18699/install.sh` to create a dev area with all the needed externals and cmssw changes.


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19735
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19729
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_1_X_2021-09-16-2300+f5f5f9/45534/triggerResults): no differences found


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000833
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000799
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-16-2300+f5f5f9/45536/triggerResults): no differences found

fwyzard commented:
OK, should be ready for review and integration.

Martin-Grunewald commented:
+1

srimanob commented:
+Upgrade

qliphy commented:
@cms-sw/pdmv-l2   Do you have any comment?

kskovpen commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
@fwyzard   Just to make sure, I see some differences on OfflinePixelPV from GPU comparisons, are they understood? :

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_1_X_2021-09-16-2300+f5f5f9/45534/10824.502_TTbar_13+2018_Patatrack_PixelOnlyGPU+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT/OfflinePixelPV_Alignment.html

fwyzard commented:
If it is in the GPU (with PR) vs GPU (without PR), yes, it is understood that the parallel processing done by the GPUs (or by the CPUs, if we were to run multithreded in the same way) can introduce some small changes, from run to run.

So, the differences are very likely not due to the PR, but to the software itself.

qliphy commented:
+1
